tensorflow/python/ops/nn_ops.py

```diff
@@ -1698,6 +1698,13 @@ def _swap_axis(logits, dim_index, last_index, name=None):
   # If dim is not the last dimension, we have to do a transpose so that we can
   # still perform softmax on its last dimension.
 
+  # In case dim is negative (and is not last dimension -1), add shape.ndims
+  if not isinstance(dim, ops.Tensor):
+    if dim < 0:
+      dim += shape.ndims
```

Comment:
Can you handle unknown shapes here? shape.ndims might be None, in which case you should use array_ops.rank